README.md
# MHWdetection

This GitHub repository contains all of the code used in the 'Detecting marine heatwaves from imperfect data' paper.   
It is set up using the package `workflowr` to ensure openness and reproducibility standards.

Please follow this link to the `workflowr` documentation website: https://robwschlegel.github.io/MHWdetection/


NEWS.md
# MHWdetection 0.0.6  (2019-03-19)

* Switching from `pkgdown` to `workflowr` for project management/visualisation

# MHWdetection 0.0.5  (2019-03-12)

* Incorporated first round of edits from AJH
* Began writing new decadal detection chunk

# MHWdetection 0.0.5  (2019-03-07)

* Incorporated first round of edits from AJS

# MHWdetection 0.0.5  (2019-03-06)

* Worked out basic event count results for all vignettes
* Final skeleton of paper

# MHWdetection 0.0.5  (2019-03-05)

* More missing data work
* Worked through basic decadal trend results

# MHWdetection 0.0.5  (2019-03-04)

* Working through the missing data analyses

# MHWdetection 0.0.5  (2019-02-26)

* Working through the duration analyses

# MHWdetection 0.0.5  (2019-02-25)

* More work on the skeleton
* Beginning to flesh out the sections

# MHWdetection 0.0.4  (2019-02-17)

* Working over the skeleton

# MHWdetection 0.0.4  (2019-02-15)

* Back to business

# MHWdetection 0.0.4  (2019-01-24)

* Looking over skeleton of main text
* Python issue was addressed
  * The difference was caused by the default nature of how each language did/did not interpolate missing data
  * The calculations otherwise still remained the same

# MHWdetection 0.0.4  (2019-01-23)

* Performed a deep dive on the issues around missing data and the quantile thresholds
    * Must look at step-by-step process in Python code to determine the issue

# MHWdetection 0.0.4  (2019-01-17)

* Checked that R and Python base results still match up after heatwaveR v0.3.6 release
    * The threshold calculations appear to be affected differently by large amounts of missing data

# MHWdetection 0.0.4  (2018-10-30)

* More editing on the time series length vignette

# MHWdetection 0.0.4  (2018-10-25)

* Big changes coming through
* Added two new vignettes:
  * `Assessing the effects of long-term trends`
  * `Assessing the effects of variance`
  

# MHWdetection 0.0.3  (2018-10-24)

* Finished editing the "simply shorter" section of the time series length vignette

# MHWdetection 0.0.3  (2018-10-23)

* Beginning aggressive editing of text
* Allowing calculation of events for all year ranges from 3 -- 34
  * Not just 10 20, and 30 years
* Removing all `var` results

# MHWdetection 0.0.3  (2018-10-14)

* Incorporated ECJO's most recent round of comments into the main two vignettes.

# MHWdetection 0.0.3  (2018-09-14)

* Added comparison to `r_vs_python.Rmd` for how the two languages handle
  missing data to create climatologies and calculate events.
* Completed the `Ice coverage` section in `missing_data.Rmd`

# MHWdetection 0.0.3  (2018-09-14)

* Finished up new proposed work on `time_series_duration.Rmd`
* Output word document version of `time_series_duration.Rmd` and `missing_data.Rmd`
* Recompiled vignettes

# MHWdetection 0.0.3  (2018-09-13)

* Finished up initially proposed work on `missing_data.Rmd`
* Rebuilt site to reflect upversion to v0.0.3

# MHWdetection 0.0.2  (2018-09-11)

* Minor additions

# MHWdetection 0.0.2  (2018-09-10)

* Further editing to `time_series_duration.Rmd`

# MHWdetection 0.0.2  (2018-09-06)

* Finished preliminary random missing data results
* Fixed CI issues with short time series vignette

# MHWdetection 0.0.2  (2018-09-05)

* Further editing to `time_series_duration.Rmd`
* Began writing `missing_data.Rmd`
* Rebuilt site to host changes

# MHWdetection 0.0.2  (2018-09-03)

* Added AJS and ECJO feedback into `MHWdetection.Rmd`
* Added AJS and ECJO feedback into `time_series_duration.Rmd`

# MHWdetection 0.0.2  (2018-06-21)

* More in-depths category comparisons made

# MHWdetection 0.0.2  (2018-06-16)

* Some pre-emptive thoughts about more category comparisons

# MHWdetection 0.0.2  (2018-06-13)

* Finished up basic comparisons of categories for different clim periods

# MHWdetection 0.0.2  (2018-06-12)

* Added significance tests for clims of different lengths for both single and re-sampled clims

# MHWdetection 0.0.2  (2018-06-07)

* More work on duration section

# MHWdetection 0.0.2  (2018-06-06)

* More work on duration section

# MHWdetection 0.0.2  (2018-06-05)

* Smoothed out issues preventing site from being built
* Tweaks to benchmarking of the languages

# MHWdetection 0.0.2  (2018-06-01)

* Second section of duration vignette nearly finished

# MHWdetection 0.0.2  (2018-05-31)

* Completed the first section of the duration vignette

# MHWdetection 0.0.2  (2018-05-30)

* Some musings added to the short duration vignette

# MHWdetection 0.0.2  (2018-05-29)

* Skeleton vignettes for the proposed research created

# MHWdetection 0.0.1  (2018-05-25)

* Regenerated vignettes to reflect corrections made in `heatwaveR` code
* Regenerated site to reflect fixes in site structure
* Added LaTeX folder 

# MHWdetection 0.0.1  (2018-05-24)

* Added a `NEWS.md` file to track changes to the package.
* Ported over the relevant vignettes that were cluttering up the `heatwaveR` site


tests/testthat.R
library(testthat)
library(MHWdetection)

test_check("MHWdetection")


tests/testthat/test-placeholder.R
context("Test placeholder.R")

test_that("placeholder doesn't fall over", {
  expect_equal(placeholder(2), 4)
  expect_is(placeholder(4), "numeric")
})


code/figures.R
# code/figures.R
# This script houses the code used for the final publication figures

source("code/functions.R")


# Figure 1 ----------------------------------------------------------------

# A multi-paneled figure showing a summary of information for the three reference time series
# These could include:
  # Time series with rug plot showing events
  # Lolliplots
  # The clims only
  # The main event
  # Map/location point
  # Summary/stats table

# Load base data
load("data/sst_ALL.Rdata")

sst_WA_event <- detect_event(ts2clm(filter(sst_ALL, site == "WA"),
                                    climatologyPeriod = c("1982-01-01", "2011-12-31")))
sst_NW_Atl_event <- detect_event(ts2clm(filter(sst_ALL, site == "NW_Atl"),
                                    climatologyPeriod = c("1982-01-01", "2011-12-31")))
sst_Med_event <- detect_event(ts2clm(filter(sst_ALL, site == "Med"),
                                    climatologyPeriod = c("1982-01-01", "2011-12-31")))

# Calculate base results
sst_ALL_res <- sst_ALL %>%
  group_by(site) %>%
  nest() %>%
  mutate(clims = map(data, ts2clm, maxPadLength = 1,
                     climatologyPeriod = c("1982-01-01", "2011-12-31")),
         events = map(clims, detect_event),
         cats = map(events, category)) %>%
  select(-data)

# Climatologies
sst_ALL_clim <- sst_ALL_res %>%
  select(-cats) %>%
  unnest(events) %>%
  filter(row_number() %% 2 == 1) %>%
  unnest(events)

# Climatologies doy
sst_ALL_clim_only <- sst_ALL_res %>%
  select(-cats) %>%
  unnest(events) %>%
  filter(row_number() %% 2 == 1) %>%
  unnest(events) %>%
  select(site, doy, seas, thresh) %>%
  unique() %>%
  arrange(site, doy)

# Events
sst_ALL_event <- sst_ALL_res %>%
  select(-cats) %>%
  unnest(events) %>%
  filter(row_number() %% 2 == 0) %>%
  unnest(events)

# Categories
sst_ALL_cat <- sst_ALL_res %>%
  select(site, cats) %>%
  unnest()

# Focus events
MHW_focus <- sst_ALL_event %>%
  group_by(site) %>%
  filter(intensity_cumulative == max(intensity_cumulative))
# Interestingly this grabs a different MHW for the Med
# so we need to more manually grab that one
focus_Med <- sst_ALL_event %>%
  filter(site == "Med", date_end <= "2004-01-01") %>%
  filter(intensity_cumulative == max(intensity_cumulative)) %>%
  left_join(sst_ALL_coords, by = "site")
focus_WA <- MHW_focus[3,] %>%
  left_join(sst_ALL_coords, by = "site")
focus_NW_Atl <- MHW_focus[2,] %>%
  left_join(sst_ALL_coords, by = "site")


# Time series with rug plot showing events
ts_clim_rug_WA <- ts_clim_rug("WA")
ts_clim_rug_WA
ts_clim_rug_NW_Atl <- ts_clim_rug("NW_Atl")
ts_clim_rug_NW_Atl
ts_clim_rug_Med <- ts_clim_rug("Med")
ts_clim_rug_Med


# Lolliplots
lolli_WA <- lolli_plot(sst_WA_event)
lolli_WA
lolli_NW_Atl <- lolli_plot(sst_NW_Atl_event)
lolli_NW_Atl
lolli_Med <- lolli_plot(sst_Med_event)
lolli_Med


# The main event
event_WA <- event_line(sst_WA_event, category = T)
event_WA
event_NW_Atl <- event_line(sst_NW_Atl_event, category = T)
event_NW_Atl
event_Med <- event_line(sst_Med_event, category = T)
event_Med


# The clims only
clim_WA <- clim_line("WA")
clim_WA
clim_NW_Atl <- clim_line("NW_Atl")
clim_NW_Atl
clim_Med <- clim_line("Med")
clim_Med


# Map/location point
map_WA <- map_point(focus_WA)
map_WA
map_NW_Atl <- map_point(focus_NW_Atl)
map_NW_Atl
map_Med <- map_point(focus_Med)
map_Med


# Summary/stats table
# Ideas for table:
  # Count of events
  # Mean/ max duration
  # mean/ max intensity
  # Seasonal range
  # Threshold range
  # Category count
table_summary_WA <- table_summary("WA")
# table_summary_WA
table_summary_NW_Atl <- table_summary("NW_Atl")
# table_summary_NW_Atl
table_summary_Med <- table_summary("Med")
# table_summary_Med


## Stitch all of the figures together
# Time series with rug plot showing events
# Lolliplots
# The clims only
# The main event
# Map/location point
# Summary/stats table
stitch_plot_WA <- ggpubr::ggarrange(ts_clim_rug_WA, clim_WA, lolli_WA, event_WA, map_WA, table_summary_WA, labels = "AUTO")
stitch_plot_WA
stitch_plot_NW_Atl <- ggpubr::ggarrange(ts_clim_rug_NW_Atl, clim_NW_Atl, lolli_NW_Atl, event_NW_Atl, map_NW_Atl, table_summary_NW_Atl, labels = "AUTO")
stitch_plot_NW_Atl
stitch_plot_Med <- ggpubr::ggarrange(ts_clim_rug_Med, clim_Med, lolli_Med, event_Med, map_Med, table_summary_Med, labels = "AUTO")
stitch_plot_Med


# Figure 2 ----------------------------------------------------------------

# A synthesis of the three tests: length, missing, trended
# The tests are all the same and the three things investigated were
  # climatologies: KS tests
load("data/sst_ALL_KS_clim.Rdata")
  # event metrics: ANOVA/Tukey OR KS tests
load("data/sst_ALL_aov_tukey.Rdata")
load("data/sst_ALL_KS_event.Rdata")
  # category count: fisher test
load("data/sst_ALL_fisher.Rdata")
# I envision here some sort of figure that compares the results side by side
# while highlighting how as the degridation of the three tests increases
# how much more rapidly this effects the results than the other tests
# Currently this is being shown with line plots but they aren't popular...

# Climatology results
sst_ALL_KS_clim_long <- sst_ALL_KS_clim %>%
  gather(key = "metric", value = "p.value", -test, -site, -rep, -index_vals) %>%
  group_by(test, site, metric, index_vals) %>%
  summarise(p.value.mean = mean(p.value))# %>%
  # mutate(index_col = paste(site, test, sep = "-"))
sst_ALL_KS_clim_long_sig <- sst_ALL_KS_clim_long %>%
  filter(p.value.mean <= 0.05)

# Event results
sst_ALL_KS_event_long <- sst_ALL_KS_event %>%
  gather(key = "metric", value = "p.value", -test, -site, -rep, -index_vals) %>%
  group_by(test, site, metric, index_vals) %>%
  summarise(p.value.mean = mean(p.value)) #%>%
# mutate(index_col = paste(test, site, metric, sep = "-"))
sst_ALL_KS_event_long_sig <- sst_ALL_KS_event_long %>%
  filter(p.value.mean <= 0.05)

# Cagtegory results
sst_ALL_fisher_long <- sst_ALL_fisher %>%
  mutate(index_vals = as.numeric(index_vals),
         metric = "category") %>%
  select(-p.adj.Fisher) %>%
  # dplyr::rename(val = p.Fisher) %>%
  # gather(key = "metric", value = "p.value", -test, -site, -rep, -index_vals) %>%
  group_by(test, site, metric, index_vals) %>%
  summarise(p.value.mean = mean(p.Fisher)) #%>%
# mutate(index_col = paste(test, site, metric, sep = "-"))
sst_ALL_fisher_long_sig <- sst_ALL_fisher_long %>%
  filter(p.value.mean <= 0.05)

# Combine for plotting
sst_ALL_plot_long <- rbind(sst_ALL_KS_clim_long, sst_ALL_KS_event_long, sst_ALL_fisher_long)
sst_ALL_plot_long_sig <- rbind(sst_ALL_KS_clim_long_sig, sst_ALL_KS_event_long_sig, sst_ALL_fisher_long_sig)

# Plot them all together
ggplot(sst_ALL_plot_long, aes(x = index_vals, y = p.value.mean, colour = metric)) +
  geom_point() +
  geom_line() +
  geom_point(data = sst_ALL_plot_long_sig, shape = 13, colour = "red") +
  # geom_errorbarh(aes(xmin = year_long, xmax = year_short)) +
  facet_grid(site~test, scales = "free_x") +
  theme(legend.position = "bottom")


# Figure 3 ----------------------------------------------------------------

# A table here showing the R2 values from Figure 2 would be good
sst_ALL_R2_long <- sst_ALL_plot_long %>%
  group_by(test, site, metric) %>%
  filter(index_vals <= 30) %>%
  nest() %>%
  mutate(R2 = map(data, lm_p_R2)) %>%
  select(-data) %>%
  unnest()


# Figure 4 ----------------------------------------------------------------

# I imagine this figure being something that supports an argument that follows
# on from one of the above results
# Specifically I think this figure will be the one that shows the average count
# of consecutive missing days depending on the percent of missing data


# Figure 5 ----------------------------------------------------------------


# Figure 6 ----------------------------------------------------------------

# A sixth figure would be a bit much, but it may be useful to show some sort
# of meta-analysis here
# Such as the relationships between increasing tests and aspects of the time series
# To that end though if such a thing is to be investiagated it would be better
# to run the global analysis and show that




code/functions.R
# code/functions.R
# This script contains all of the functions used throughout 'code/workflow.R'


# Libraries ---------------------------------------------------------------

library(tidyverse)
library(broom)
library(heatwaveR, lib.loc = "~/R-packages/")
# cat(paste0("heatwaveR version = ",packageDescription("heatwaveR")$Version))
library(lubridate) # This is intentionally activated after data.table
library(fasttime)
library(ggpubr)
library(boot)
library(FNN)
library(mgcv)
library(doMC); doMC::registerDoMC(cores = 50)
library(ggridges)
library(ggpubr)
# library(rcompanion)


# Meta-data ---------------------------------------------------------------

MHW_colours <- c(
  "I Moderate" = "#ffc866",
  "II Strong" = "#ff6900",
  "III Severe" = "#9e0000",
  "IV Extreme" = "#2d0000"
)

OISST_files <- dir(path = "~/data/OISST", full.names = T)

category_files <- as.character(dir(path = "~/data/cat_clim", pattern = "cat.clim",
                                   full.names = TRUE, recursive = TRUE))

sst_ALL_coords <- data.frame(site = c("WA", "NW_Atl", "Med"),
                             lon = c(112.5, -67, 9),
                             lat = c(-29.5, 43, 43.5))


# Re-sampling -------------------------------------------------------------

# Function for creating a re-sample from 37 years of data
sample_37 <- function(rep){
  year_filter <- c(1,2,3)
  while(length(unique(year_filter)) != 37){
    year_filter <- factor(sample(1982:2018, size = 37, replace = F))
  }
  res <- data.frame()
  for(i in 1:length(year_filter)){
    df_sub <- sst_ALL %>%
      filter(year(t) == year_filter[i],
             # Remove leap-days here
             paste0(month(t),"-",day(t)) != "2-29") %>%
      mutate(year = seq(1982,2018)[i],
             month = month(t),
             day = day(t),
             year_orig = year(t)) %>%
      mutate(t = as.Date(fastPOSIXct(paste(year, month, day, sep = "-")))) %>%
      select(-year, -month, -day)
    res <- rbind(res, df_sub)
  }
  res$rep <- as.character(rep)
  return(res)
}


# De-trending -------------------------------------------------------------

detrend <- function(df){
  resids <- broom::augment(lm(temp ~ t, df))
  res <- df %>%
    mutate(temp = temp - resids$.fitted) %>%
    select(-year_orig)
  return(res)
}


# Random knockout ---------------------------------------------------------

# Function for knocking out data but maintaing the time series consistency
random_knockout <- function(prop, df = sst_ALL_flat){
  # NB: Don't allow samppling of first and last value to ensure
  # all time series are the same length
  ts_length <- nrow(filter(df, site == "WA"))
  miss_index <- sample(seq(2, ts_length-1, 1), ts_length*prop, replace = F)
  # df$x1[sample(nrow(df),250)] <- NA
  res <- df %>%
    group_by(site, rep) %>%
    mutate(row_index = 1:n(),
           temp = replace(temp, which(row_index %in% miss_index), NA)) %>%
    mutate(index_vals = prop) %>%
    select(-row_index)
  return(res)
}


# Count consecutive days --------------------------------------------------

# Quantify consecutive missing days
con_miss <- function(df){
  ex1 <- rle(is.na(df$temp))
  ind1 <- rep(seq_along(ex1$lengths), ex1$lengths)
  s1 <- split(1:nrow(df), ind1)
  res <- do.call(rbind,
                 lapply(s1[ex1$values == TRUE], function(x)
                   data.frame(index_start = min(x), index_end = max(x))
                 )) %>%
    mutate(duration = index_end - index_start + 1) %>%
    group_by(duration) %>%
    summarise(count = n())
  return(res)
}


# Add trends --------------------------------------------------------------

# Function for adding trends to data
add_trend <- function(rate, df = sst_ALL_flat){
  daily_step <- rate/3652.5
  res <- df %>%
    group_by(site, rep) %>%
    mutate(row_index = 1:n(),
           temp = temp + (row_index*daily_step)) %>%
    mutate(index_vals = rate) %>%
    select(-row_index)
  return(res)
}


# Calculate clims, events, and cats ---------------------------------------

# The climatologies themselves are much smaller and easier to handle on their own
# So we want to pull out only the 366 day clims per run
clim_only <- function(df){
  res <- df$climatology %>%
    select(doy, seas:thresh) %>%
    unique() %>%
    mutate_if(is.numeric, round, 3) %>%
    arrange(doy)
  return(res)
}

# Likewise we want to grab only the event values we are interested in
event_only <- function(df){
  res <- df$event %>%
    select(date_start, date_peak, date_end, duration, intensity_mean, intensity_max, intensity_cumulative) %>%
    mutate_if(is.numeric, round, 3)
  return(res)
}

# Calculate climatologies, events, and categories on full length time series
# Set `maxPadLength = 1` so that ts2clm() will fill in the missing leap year days
# caused by the re-sampling step above
# testers...
# df <- sst_ALL_knockout %>%
  # filter(site == "WA", rep == "1", index_vals == 0)
clim_event_cat_calc <- function(df){
  res <- df %>%
    nest() %>%
    mutate(clims = map(data, ts2clm, maxPadLength = 1,
                       climatologyPeriod = c("1982-01-01", "2011-12-31")),
           events = map(clims, detect_event),
           cat = map(events, category),
           clim = map(events, clim_only),
           event = map(events, event_only)) %>%
    select(-data, -clims, -events)
  return(res)
}


# Calculate clim, events, and cats on short time series -------------------

# NB: The shortened time series require their own function for calculating results
# in order to match the length of any given time series
# testers...
# df <- sst_ALL_flat %>%
  # filter(site == "WA", rep == "1")
# year_begin <- 2000
shrinking_results <- function(year_begin, df = sst_ALL_flat){
  res <- df %>%
    filter(year(t) >= year_begin) %>%
    mutate(index_vals = 2018-year_begin+1) %>%
    group_by(site, rep, index_vals) %>%
    nest() %>%
    mutate(clims = map(data, ts2clm, maxPadLength = 1,
                       climatologyPeriod = c(paste0(year_begin,"-01-01"), "2018-12-31")),
           events = map(clims, detect_event),
           cat = map(events, category),
           clim = map(events, clim_only),
           event = map(events, event_only)) %>%
    select(-data, -clims, -events)
  return(res)
}


# KS tests ----------------------------------------------------------------

# The problem with running KS tests on category data is that it doesn't appear
# to be bothered by differences in sample sizes
# Meaning that a comparison across the count of events matters more on what
# counts are present, rather than the difference in total counts

# Function for running a pairwise KS test against the index_standard
# Lazily I have it run on chosen columns by name given certain tests
# as getting it to do this programmatically was becoming annoying
# testers...
# df_2 <- filter(res, index_vals == 20)
# df_1 <- df_standard
KS_sub <- function(df_2, df_1){
  if("seas" %in% names(df_1)){
    suppressWarnings( # Suppress warnings about ties
    res <- data.frame(seas = round(ks.test(df_1$seas, df_2$seas)$p.value, 4),
                      thresh = round(ks.test(df_1$thresh, df_2$thresh)$p.value, 4))
    )
  }
  if("intensity_mean" %in% names(df_1)){
    suppressWarnings( # Suppress warnings about ties
      res <- data.frame(duration = round(ks.test(df_1$duration, df_2$duration)$p.value, 4),
                        intensity_mean = round(ks.test(df_1$intensity_mean, df_2$intensity_mean)$p.value, 4),
                        intensity_max = round(ks.test(df_1$intensity_max, df_2$intensity_max)$p.value, 4),
                        intensity_cumulative = round(ks.test(df_1$intensity_cumulative, df_2$intensity_cumulative)$p.value, 4))
    )
  }
  if("category" %in% names(df_1)){
    df_1$category <- factor(df_1$category, levels = c("I Moderate", "II Strong", "III Severe", "IV Extreme"))
    df_2$category <- factor(df_2$category, levels = c("I Moderate", "II Strong", "III Severe", "IV Extreme"))
    suppressWarnings( # Suppress warnings about ties
      res <- data.frame(category = round(ks.test(as.numeric(df_1$category), as.numeric(df_2$category))$p.value, 4))
    )
  }
  return(res)
}

# Wrapper function to run all pairwise KS tests
# df <- sst_ALL_clim_event_cat[ ,c(1:4,6)] %>%
#   filter(test == "missing", site == "WA", rep == "1") %>%
#   select(-test, -rep, -site)
KS_p <- function(df){
  # Unnest the clim data
  suppressWarnings( # Suppress warning about different factor levels for cat data
  df_long <- df %>%
    # select(-event, -cat) %>%
    unnest()
  )
  # This determines if the data are from the length test or not
  # It then sets the benchmark value (30 years length or 0% missing/ 0C-dec)
  if(max(df_long$index_vals) > 1){
    index_filter <- 30
    if("intensity_mean" %in% names(df_long)){
      df_long <- df_long %>%
        filter(index_vals >= 10,
               date_start >= "2009-01-02", date_end <= "2018-12-31")
    }
    if("category" %in% names(df_long)){
      df_long <- df_long %>%
        filter(index_vals >= 10,
               peak_date >= "2009-01-01", peak_date <= "2018-12-31")
    }
  } else {
    index_filter <- 0
  }
  df_standard <- df_long %>%
    filter(index_vals == index_filter)
  res <- df_long %>%
    filter(index_vals != index_filter) %>%
    group_by(index_vals) %>%
    nest() %>%
    mutate(KS = map(data, KS_sub, df_1 = df_standard)) %>%
    select(-data) %>%
    unnest()
  return(res)
}


# AOV and Tukey tests -----------------------------------------------------

# Run an ANOVA on each metric of the combined event results and get the p-value
# df <- res
aov_p <- function(df){
  aov_models <- df[ , -grep("index_vals", names(df))] %>%
    map(~ aov(.x ~ df$index_vals)) %>%
    map_dfr(~ broom::tidy(.), .id = 'metric') %>%
    mutate(p.value = round(p.value, 4)) %>%
    filter(term != "Residuals") %>%
    select(metric, p.value)
  return(aov_models)
}

# Run an ANOVA on each metric and then a Tukey test
tukey_p <- function(df){
  tukey_models <- df[ , -grep("index_vals", names(df))] %>%
    map(~ TukeyHSD(aov(.x ~ df$index_vals))) %>%
    map_dfr(~ broom::tidy(.), .id = 'metric') %>%
    mutate(adj.p.value = round(adj.p.value, 4)) %>%
    select(metric, comparison, adj.p.value) %>%
    arrange(metric, adj.p.value)
  return(tukey_models)
}

# Quick wrapper for getting results for ANOVA and Tukey on clims
# df <- sst_ALL_clim_event_cat %>%
  # filter(test == "length", site == "WA", rep == "1") %>%
  # select(-test, -rep, -site)
aov_tukey <- function(df){
  # Extract MHW event data
  prep <- df %>%
    select(index_vals, event) %>%
    unnest()
  # Filter out results not from the most recent decade for equal comparison
  # when looking at the changing lengtths test
  if(max(prep$index_vals) > 1){
    prep <- prep %>%
      filter(index_vals >= 10) %>%
      # 2009-01-02 is intentional
      filter(date_start >= "2009-01-02", date_end <= "2018-12-31")
  }
  # Calculate the ANOVA and Tukey p values
  res <- prep %>%
    select(-date_start, -date_peak, -date_end) %>%
    mutate(index_vals = as.factor(as.character(index_vals))) %>%
    nest() %>%
    mutate(aov = map(data, aov_p),
           tukey = map(data, tukey_p)) %>%
    select(-data)
  return(res)
}



# Fisher test for category counting ---------------------------------------


# Fisher pairwise function
# Takes one rep of missing data and compares it against the complete data
# testers...
# df <- unnest(slice(select(res, data), 1))
# df_comp <- df_standard
fisher_pair <- function(df, df_comp){
  # Prep data
  df_joint <- table(rbind(df_comp, df))
  if(ncol(df_joint) == 1){
    df_joint <- as.table(cbind(df_joint, 'III & IV' = c(0,0)))
  }
  # Run tests
  # res <- round(fisher.test(df_joint)$p.value, 4)
  res <- fisher.test(df_joint)
  # res_broom <- broom::augment(res) #%>%
    # mutate_if(is.numeric, round, 4)
  return(res)
}

# Wrapper to get data ready for pair-wise chi-squared tests
# testers...
# df <- sst_ALL_clim_event_cat %>%
  # filter(test == "length", site == "WA", rep == "1") %>%
  # select(-test, -rep, -site)
fisher_test <- function(df){
  suppressWarnings( # Suppress warning about category levels not all being present
    df_long <- df %>%
      select(index_vals, cat) %>%
      unnest() %>%
      # select(index_vals, category) %>%
      mutate(category = ifelse(category %in% c("I Moderate","II Strong"), "I & II", category),
             category = ifelse(category %in% c("III Severe","IV Extreme"), "III & IV", category),
             category = as.character(category))
  )
  # This determines if the data are from the length test or not
  # It then sets the benchmark value (30 years length or 0% missing/ 0C-dec)
  if(max(df_long$index_vals) > 1){
    index_filter <- 30
    df_long <- df_long %>%
      filter(index_vals >= 10,
             peak_date >= "2009-01-01", peak_date <= "2018-12-31")
  } else {
    index_filter <- 0
  }
  res_table <- table(select(df_long, index_vals, category))
  if(ncol(res_table) == 1){
    res_table <- as.table(cbind(res_table, 'III & IV' = rep(0, nrow(res_table))))
  }
  res <- rcompanion::pairwiseNominalIndependence(res_table, fisher = TRUE, gtest = FALSE,
                                     chisq = FALSE, method = "bonferroni",
                                     digits = 3)
  res_tidy <- separate(res, col = "Comparison", into = c("group_1", "group_2"), sep = " : ") %>%
    filter(group_1 == index_filter | group_2 == index_filter) %>%
    mutate(index_vals = ifelse(group_1 == index_filter, group_2, group_1),
           p.adj.Fisher = p.Fisher* n(),
           p.adj.Fisher = ifelse(p.adj.Fisher > 1, 1, p.adj.Fisher)) %>%
    select(index_vals, p.Fisher, p.adj.Fisher)
  return(res_tidy)
}


# Model linear relationships ----------------------------------------------

# Wrapper function for nesting
# testers...
# df <- slice(sst_ALL_R2_long, 1) %>%
  # unnest()
lm_p_R2 <- function(df){
  res <- lm(p.value.mean~index_vals, data = df)
  res_broom <- broom::glance(res)
  return(res_broom)
}


# Figure convenience functions --------------------------------------------

# Expects a one row data.frame with a 'lon' and 'lat' column
# df <- focus_WA
map_point <- function(df){
  category_data <- readRDS(category_files[grepl(pattern = as.character(df$date_peak),
                                                x = category_files)])
  map_out <- ggplot(data = df, aes(x = lon, y = lat)) +
    geom_tile(data = category_data, aes(fill = category)) +
    borders(fill = "grey80", colour = "black") +
    geom_point(shape = 21, colour = "white", fill = "hotpink", size = 2) +
    scale_fill_manual("Category",
                      values = c("#ffc866", "#ff6900", "#9e0000", "#2d0000"),
                      labels = c("I Moderate", "II Strong",
                                 "III Severe", "IV Extreme")) +
    coord_cartesian(xlim = c(df$lon[1]-20, df$lon[1]+20),
                    ylim = c(df$lat[1]-20, df$lat[1]+20)) +
    labs(x = NULL, y = NULL) +
    theme(legend.position = "bottom")
  return(map_out)
}

# This function expects the output of the clim, event, cat pipe
# df <- filter(sst_ALL_res, site == "NW_Atl")
table_summary <- function(site_1){

  df <- filter(sst_ALL_res, site == site_1)

  # Seasonal min/mean/max
  # Threshold min/mean/max
  clim_long <- df %>%
    select(clims) %>%
    unnest() %>%
    select(doy, seas, thresh) %>%
    unique() %>%
    select(seas, thresh) %>%
    gather(key = "metric", value = "val")

  event_long <- df %>%
    select(events) %>%
    unnest() %>%
    filter(row_number() %% 2 == 0) %>%
    unnest() %>%
    select(duration, intensity_mean, intensity_max, intensity_cumulative) %>%
    gather(key = "metric", value = "val")

  event_count <- df %>%
    select(cats) %>%
    unnest() %>%
    select(category) %>%
    group_by(category) %>%
    summarise(val = n()) %>%
    mutate(metric = "count") %>%
    spread(key = category, value = val)
  if(!"IV Extreme" %in% colnames(event_count)){
    event_count <- cbind(event_count, tibble('IV Extreme' = 0))
  }
  event_count <- event_count %>%
    select(metric, 'I Moderate', 'II Strong', 'III Severe', 'IV Extreme') %>%
    dplyr::rename(' ' = metric)

  summary_res <- rbind(clim_long, event_long) %>%
    group_by(metric) %>%
    summarise_all(.funs = c("min", "mean", "max", "sd")) %>%
    mutate_if(is.numeric, round, 2) %>%
    arrange(metric)

  tbl_1 <- gridExtra::tableGrob(summary_res, rows = NULL)
  tbl_2 <- gridExtra::tableGrob(event_count, rows = NULL)

  tbl_all <- gridExtra::grid.arrange(tbl_1, tbl_2,
                                     nrow = 2, as.table = TRUE)

  return(tbl_all)
  # res_list <- list(summary_res = summary_res,
                   # event_count = event_count)
  # return(res_list)
}

# Wrapper for time series + clims + event rug
ts_clim_rug <- function(site_1){
  ts_data <- filter(sst_ALL, site == site_1)
  clim_data <- filter(sst_ALL_clim, site == site_1)
  event_data <- filter(sst_ALL_event, site == site_1)
  fig <- ggplot(data = ts_data, aes(x = t, y = temp)) +
    geom_line(colour = "grey20") +
    geom_line(data = clim_data, aes(y = seas),
              linetype = "dashed", colour = "steelblue3") +
    geom_line(data = clim_data, linetype = "dotted", colour = "tomato3",
              aes(x = t, y = thresh)) +
    geom_rug(data = event_data, sides = "b", colour = "red3", size = 2,
             aes(x = date_peak, y = min(ts_data$temp))) +
    labs(x = NULL, y = "Temperature (°C)")
  return(fig)
}

# Wrapper for clim only line plot
clim_line <-function(site_1){
  clim_data <- filter(sst_ALL_clim_only, site == site_1)
  ggplot(data = clim_data, aes(x = doy)) +
    geom_line(aes(y = seas), colour = "steelblue3") +
    geom_line(aes(y = thresh), colour = "tomato3") +
    labs(x = NULL, y = "Temperature (°C)")
}


# Global functions --------------------------------------------------------

# The following wrapper functions use the above functions,
# but allow them to be used on the NOAA OISST NetCDF file structure
